hugr/src/algorithm/merge_bbs.rs

```diff
@@ -0,0 +1,245 @@
+        Output {
+            types: succ_sig.output.clone(),
+        },
+    );
+
+    let dfg1 = replacement.add_node_with_parent(
+        merged,
+        DFG {
+            signature: pred_ty.inner_signature().clone(),
+        },
+    );
+    for (i, _) in pred_ty.inputs.iter().enumerate() {
+        replacement.connect(input, i, dfg1, i)
+    }
+
+    let dfg2 = replacement.add_node_with_parent(
+        merged,
+        DFG {
+            signature: succ_sig.clone(),
+        },
+    );
+    for (i, _) in succ_sig.output.into_iter().enumerate() {
+        replacement.connect(dfg2, i, output, i)
+    }
+
+    // At the junction, must unpack the first (tuple, branch predicate) output
+    let tuple_elems = pred_ty.sum_rows.clone().into_iter().exactly_one().unwrap();
+    let unp = replacement.add_node_with_parent(
+        merged,
+        UnpackTuple {
+            tys: tuple_elems.clone(),
+        },
+    );
+    replacement.connect(dfg1, 0, unp, 0);
+    let other_start = tuple_elems.len();
+    for (i, _) in tuple_elems.into_iter().enumerate() {
+        replacement.connect(unp, i, dfg2, i)
+    }
+    for (i, _) in pred_ty.other_outputs.into_iter().enumerate() {
+        replacement.connect(dfg1, i + 1, dfg2, i + other_start)
+    }
+
+    let rep = Replacement {
+        removal: vec![pred, succ],
+        replacement,
+        adoptions: HashMap::from([(dfg1, pred), (dfg2, succ)]),
+        mu_inp: cfg
+            .all_linked_outputs(pred)
+            .map(|(src, src_pos)| NewEdgeSpec {
+                src,
+                tgt: merged,
+                kind: NewEdgeKind::ControlFlow { src_pos },
+            })
+            .collect(),
+        mu_out: cfg
+            .node_outputs(succ)
+            .map(|src_pos| NewEdgeSpec {
+                src: merged,
+                tgt: cfg
+                    .linked_inputs(succ, src_pos)
+                    .exactly_one()
+                    .ok()
+                    .unwrap()
+                    .0,
+                kind: NewEdgeKind::ControlFlow { src_pos },
+            })
+            .collect(),
+        mu_new: vec![],
+    };
+    (rep, dfg1, dfg2)
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::HashSet;
+
+    use itertools::Itertools;
+
+    use crate::builder::{CFGBuilder, Dataflow, HugrBuilder};
+    use crate::extension::prelude::{QB_T, USIZE_T};
+    use crate::extension::{ExtensionRegistry, ExtensionSet, PRELUDE, PRELUDE_REGISTRY};
+    use crate::hugr::views::sibling::SiblingMut;
+    use crate::ops::handle::CfgID;
+    use crate::ops::{Const, Noop, OpType};
+    use crate::types::{FunctionType, Type, TypeRow};
+    use crate::{const_extension_ids, type_row, Extension, HugrView};
+
+    use super::merge_basic_blocks;
+
+    const_extension_ids! {
+        const EXT_ID: ExtensionId = "TestExt";
+    }
+    #[test]
+    fn in_loop() -> Result<(), Box<dyn std::error::Error>> {
+        /* -> Noop1 -----> Test -> Exit       -> Noop1AndTest --> Exit
```

Comment:
I feel this shape is a reasonable test, so am not tooo worried about adding more control-flow structures. However it'd definitely be good to add more differently-*typed* blocks e.g. with different numbers of inputs/outputs, values in the branch-control predicate or not, etc.